spec/tom_queue/worker_supervisor_spec.rb

```diff
@@ -0,0 +1,109 @@
+require 'tom_queue/worker_supervisor'
+
+describe TomQueue::WorkerSupervisor do
+  let(:supervisor) { described_class.new }
+  describe "#supervise" do
+    it "creates a process with the right name" do
+      supervisor.supervise(as: "worker") { }
+      expect(supervisor.processes.keys).to include "worker1"
+    end
+
+    it "creates a process with the specified behaviour" do
+      my_proc = -> {}
+      supervisor.supervise(as: "worker", &my_proc)
+      expect(supervisor.processes.values).to include my_proc
+    end
+
+    it "can create multiple processes" do
+      supervisor.supervise(as: "worker", count: 2) { }
+      supervisor.supervise(as: "deferred_worker") { }
+      expect(supervisor.processes.keys).to match_array ["worker1", "worker2", "deferred_worker1"]
+    end
+  end
+
+  describe "#before_fork" do
+    it "assigns the before_fork task" do
+      before_fork_task = -> { }
+      supervisor.before_fork = before_fork_task
+      expect(supervisor.before_fork).to eq before_fork_task
+    end
+  end
+
+  describe "#after_fork" do
+    it "assigns the after_fork task" do
+      after_fork_task = -> { }
+      supervisor.after_fork = after_fork_task
+      expect(supervisor.after_fork).to eq after_fork_task
+    end
+  end
+
+  describe "#run" do
+    let(:process) { -> { p "executing process" } }
+    let(:process_id) { 1 }
+    before { supervisor.supervise(as: "worker", &process) }
+
+    before do
+      # we just want the loop to iterate once for testing
+      allow(supervisor).to receive(:stop_loop?).and_return(true)
+      # we don't want to sleep in the tests, so we stub the throttle method
+      allow(supervisor).to receive(:throttle_loop)
+      # we don't want to wait for anything in the test, so we're pretending the child process exits
+      allow(supervisor).to receive(:wait_for_child_process_to_exit).and_return([1, "worker1"])
+      # we're stubbing fork so we don't create a new process in the test
+      allow(supervisor).to receive(:fork) do |&block|
+        block.call
+        process_id
+      end
```

Comment:
Lots of stubbing here to get around forking/sleeping in the tests.

Reply:
I'd be tempted to _actually_ spawn a child process in the tests and assert behaviours

Reply:
heh, I'll timebox it this morning and see where I get!